src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -911,4 +1158,192 @@ public sealed override int GetHashCode()
             throw Roslyn.Utilities.ExceptionUtilities.Unreachable();
         }
     }
+
+    /// <summary>
+    /// A helper method used in static constructor of <see cref="DataSectionStringType"/>.
+    /// to share IL and hence save assembly size.
+    /// </summary>
+    file sealed class BytesToStringHelper(
+        Cci.INamespaceTypeDefinition containingType,
+        Cci.IMethodReference encodingGetString,
+        ushort maxStack,
+        ImmutableArray<byte> il)
+        : DefaultMethodDef(
+            PrivateImplementationDetails.SynthesizedBytesToStringFunctionName,
+            Cci.TypeMemberVisibility.Private,
+            containingType,
+            maxStack,
+            il)
+    {
+        private readonly ImmutableArray<Cci.IParameterDefinition> _parameters =
+        [
+            new BytesParameter(encodingGetString),                              // byte* bytes
+            new ParameterDefinition(1, "length", Cci.PlatformType.SystemInt32), // int length
+        ];
+
+        public override ImmutableArray<Cci.IParameterDefinition> Parameters => _parameters;
+        public override Cci.ITypeReference GetType(EmitContext context) => context.Module.GetPlatformType(Cci.PlatformType.SystemString, context);
+
+        private sealed class BytesParameter(
+            Cci.IMethodReference encodingGetString)
+            : ParameterDefinitionBase(0, "bytes")
+        {
+            private readonly Cci.IMethodReference _encodingGetString = encodingGetString;
+
+            public override Cci.ITypeReference GetType(EmitContext context)
+            {
+                return _encodingGetString.GetParameters(context)[0].GetType(context);
+            }
+        }
+    }
+
+    file sealed class StaticConstructor(
+        Cci.INamespaceTypeDefinition containingType,
+        ushort maxStack,
+        ImmutableArray<byte> il)
+        : DefaultMethodDef(
+            WellKnownMemberNames.StaticConstructorName,
+            Cci.TypeMemberVisibility.Private,
+            containingType,
+            maxStack,
+            il)
+    {
+        public override bool IsRuntimeSpecial => true;
+        public override bool IsSpecialName => true;
+    }
+
+    file class DefaultMethodDef(
```

Comment:
>DefaultMethodDef [](http://example.com/codeflow?start=15&length=16)

I do not think we have a concept of "default method definition" and I think it would be better to not introduce one

Reply:
It looks like we already have ```RootModuleStaticConstructor``` that implements ```Cci.IMethodDefinition, Cci.IMethodBody```. Let's extract a common abstract base "MethodDefinition", rename ```RootModuleStaticConstructor``` to ```StaticConstructor``` and reuse the new base across ```StaticConstructor``` and ```BytesToStringHelper```.

Reply:
While extracting common bases, I recommend extract/copy/paste code from pre-existing types, without changing relative order of members. That is going to make code review much easier.